daml-foundations/daml-ghc/damldoc/src/DA/Daml/GHC/Damldoc/Render/Rst.hs

```diff
@@ -150,71 +155,44 @@ fieldTable fds = T.unlines $ -- NB final empty line is essential and intended
-         , prefix "  - " $ maybe " " (markdownToRst . T.unwords . T.lines) fd_descr ]
+         , prefix "  - " $ maybe " " (docTextToRst . DocText . T.unwords . T.lines . unDocText) fd_descr ] -- FIXME: this makes no sense
        | FieldDoc{..} <- fds ]
 
-
--- | Render fields as a table, like this:
--- >  ============ ==============================
--- >  **Field**    **Type/Description**
--- >  ============ ==============================
--- >  **anA**      | *a*
--- >  **another**  | *a*
--- >               | another a
--- >  **andText**  | *Text*
--- >               | and text
--- >  ============ ==============================
-_fieldTableVerbatim :: [FieldDoc] -> T.Text
-_fieldTableVerbatim []  = "(no fields)"
-_fieldTableVerbatim fds = header <> fieldRows <> equalLines
-  where
-    equalLines = T.unwords (map (flip T.replicate "-") [fLen, 30])
-                 -- 30 because last overline can be shorter
-    fLen = 4 + (maximum $ map T.length $ fieldHdr : map fd_name fds)
-           -- +4 for field names in bold face
-    fieldHdr = "Field"
-    header = T.unlines
-      [ equalLines
-      , adjust fLen (enclosedIn "**" fieldHdr) <> " " <> "**Type/Description**"
-      , equalLines
-      ]
-    fieldRows = T.unlines
-      [ adjust fLen (enclosedIn "**" fd_name) <> " | "
-        <> enclosedIn "*" (type2rst fd_type)
-        <> maybe "" (\d -> "\n" <> indent (fLen + 1) (prefix "| " (markdownToRst d))) fd_descr
-      | FieldDoc{..} <- fds ]
-
 -- | Render a type. Nested type applications are put in parentheses.
 type2rst :: Type -> T.Text
 type2rst = f (0 :: Int)
   where
     -- 0 = no brackets
     -- 1 = brackets around function
     -- 2 = brackets around function AND application
-    f _ (TypeApp n []) = n
-    f i (TypeApp n as) = (if i >= 2 then inParens else id) $ T.unwords (n : map (f 2) as)
+    f _ (TypeApp n []) = unTypename n
+    f i (TypeApp n as) = (if i >= 2 then inParens else id) $ T.unwords (unTypename n : map (f 2) as)
     f i (TypeFun ts) = (if i >= 1 then inParens else id) $ T.intercalate " -> " $ map (f 1) ts
     f _ (TypeList t1) = "[" <> f 0 t1 <> "]"
     f _ (TypeTuple ts) = "(" <> T.intercalate ", " (map (f 0) ts) <>  ")"
 
 
 fct2rst :: Modulename -> FunctionDoc -> T.Text
-fct2rst md_name  FunctionDoc{..} =
-  renderAnchor (functionAnchor md_name fct_name fct_type) <> "\n"
-  <> enclosedIn "**" (if isAlpha (T.head fct_name) then fct_name else inParens fct_name) <> "\n  : "
-  <> maybe "" ((<> " => ") . type2rst) fct_context
-  <> maybe "" ((<> "\n\n") . type2rst) fct_type
-  <> maybe "" (indent 2 . markdownToRst) fct_descr
-  <> "\n"
+fct2rst md_name  FunctionDoc{..} = T.unlines
+    [ renderAnchor (functionAnchor md_name fct_name fct_type)
+    , enclosedIn "**" (wrapOp (unFieldname fct_name))
+    , T.concat
+        [ "  : "
+        , maybe "" ((<> " => ") . type2rst) fct_context
+        , maybe "" ((<> "\n\n") . type2rst) fct_type
+            -- FIXME: when would a function not have a type?
```

Comment:
Functions could have no (explicitly-stated) type but a doc comment.
(as long as we still use the _parser_ AST... the type checker will give every function a type)

Reply:
Ah, makes sense! We should probably switch to typechecker AST like you were saying.

Reply:
Maybe worth replacing the `FIXME`s by explanations?